File: c++/medianOfTwoSortedArrays.cpp
/**
 * @file medianOfTwoSortedArrays.cpp
 * @brief Solution to find the median of two sorted arrays in O(log(min(m,n))) time
 * @author Hacktoberfest Contributor
 * @date October 2025
 * 
 * Problem: Given two sorted arrays nums1 and nums2 of size m and n respectively,
 * return the median of the two sorted arrays.
 * 
 * Algorithm: Binary Search Approach
 * - Use binary search on the smaller array to find the correct partition
 * - Ensure that elements on the left side are smaller than elements on the right side
 * - The median is calculated based on the maximum of left elements and minimum of right elements
 * 
 * Time Complexity: O(log(min(m, n)))
 * Space Complexity: O(1)
 */

#include <vector>
#include <algorithm>
#include <climits>
#include <iostream>
using namespace std;

class Solution {
public:
    /**
     * @brief Finds the median of two sorted arrays using binary search
     * @param nums1 First sorted array
     * @param nums2 Second sorted array
     * @return The median value as a double
     */
    double findMedianSortedArrays(vector<int>& nums1, vector<int>& nums2) {
        // Ensure nums1 is the smaller array for optimization
        // This reduces the search space in binary search
        if (nums1.size() > nums2.size()) {
            return findMedianSortedArrays(nums2, nums1);
        }

        int m = nums1.size();  // Size of first array
        int n = nums2.size();  // Size of second array
        int left = 0;          // Left boundary for binary search
        int right = m;         // Right boundary for binary search

        // Binary search to find the correct partition
        while (left <= right) {
            // Partition point for nums1
            int partitionA = (left + right) / 2;
            
            // Partition point for nums2 
            // Total left elements should be (m+n+1)/2
            int partitionB = (m + n + 1) / 2 - partitionA;

            // Find boundary elements around the partition
            // Maximum element on left side of nums1
            int maxLeftA = (partitionA == 0) ? INT_MIN : nums1[partitionA - 1];
            
            // Minimum element on right side of nums1
            int minRightA = (partitionA == m) ? INT_MAX : nums1[partitionA];
            
            // Maximum element on left side of nums2
            int maxLeftB = (partitionB == 0) ? INT_MIN : nums2[partitionB - 1];
            
            // Minimum element on right side of nums2
            int minRightB = (partitionB == n) ? INT_MAX : nums2[partitionB];

            // Check if we found the correct partition
            // All elements on left should be <= all elements on right
            if (maxLeftA <= minRightB && maxLeftB <= minRightA) {
                // Found the correct partition
                
                // If total elements are even, median is average of middle two elements
                if ((m + n) % 2 == 0) {
                    return (max(maxLeftA, maxLeftB) + min(minRightA, minRightB)) / 2.0;
                } 
                // If total elements are odd, median is the maximum of left elements
                else {
                    return max(maxLeftA, maxLeftB);
                }
            } 
            // If maxLeftA > minRightB, we need to move left in nums1
            else if (maxLeftA > minRightB) {
                right = partitionA - 1;
            } 
            // If maxLeftB > minRightA, we need to move right in nums1
            else {
                left = partitionA + 1;
            }
        }

        // This should never be reached if input arrays are valid
        return 0.0;
    }
};

/**
 * @brief Example usage and test cases
 */
int main() {
    Solution solution;
    
    // Test case 1: Even total elements
    vector<int> nums1 = {1, 3};
    vector<int> nums2 = {2};
    cout << "Test 1 - Arrays: [1,3] and [2]" << endl;
    cout << "Median: " << solution.findMedianSortedArrays(nums1, nums2) << endl;
    
    // Test case 2: Odd total elements
    nums1 = {1, 2};
    nums2 = {3, 4};
    cout << "\nTest 2 - Arrays: [1,2] and [3,4]" << endl;
    cout << "Median: " << solution.findMedianSortedArrays(nums1, nums2) << endl;
    
    return 0;
}

File: python/Median_of_Two_Sorted_Arrays.py
class Solution:
    def findMedianSortedArrays(self, nums1, nums2):
        if len(nums1) > len(nums2):
            nums1, nums2 = nums2, nums1
        
        m, n = len(nums1), len(nums2)
        
        left, right = 0, m
        
        while left <= right:
            partition1 = (left + right) // 2
            
            partition2 = (m + n + 1) // 2 - partition1
            
            max_left1 = float('-inf') if partition1 == 0 else nums1[partition1 - 1]
            max_left2 = float('-inf') if partition2 == 0 else nums2[partition2 - 1]
            
            min_right1 = float('inf') if partition1 == m else nums1[partition1]
            min_right2 = float('inf') if partition2 == n else nums2[partition2]
            
            if max_left1 <= min_right2 and max_left2 <= min_right1:
                if (m + n) % 2 == 0:
                    return (max(max_left1, max_left2) + min(min_right1, min_right2)) / 2.0
                else:
                    return max(max_left1, max_left2)
            
            elif max_left1 > min_right2:
                right = partition1 - 1
            else:
                left = partition1 + 1
        
        raise ValueError("Input arrays are not sorted")

    def findMedianSortedArrays_bruteforce(self, nums1, nums2):
        merged = []
        i = j = 0
        
        while i < len(nums1) and j < len(nums2):
            if nums1[i] <= nums2[j]:
                merged.append(nums1[i])
                i += 1
            else:
                merged.append(nums2[j])
                j += 1
        
        while i < len(nums1):
            merged.append(nums1[i])
            i += 1
        
        while j < len(nums2):
            merged.append(nums2[j])
            j += 1
        
        n = len(merged)
        if n % 2 == 0:
            return (merged[n // 2 - 1] + merged[n // 2]) / 2.0
        else:
            return merged[n // 2]


def test_solution():
    solution = Solution()
    
    test_cases = [
        ([1, 3], [2], 2.0),
        ([1, 2], [3, 4], 2.5),
        ([0, 0], [0, 0], 0.0),
        ([], [1], 1.0),
        ([2], [], 2.0),
        ([1, 3, 8], [7, 9, 10, 11], 8.0),
        ([1, 2, 3, 4], [5, 6, 7, 8], 4.5),
        ([1], [2, 3, 4, 5, 6], 3.5)
    ]
    
    print("Testing Median of Two Sorted Arrays:")
    print("=" * 50)
    
    for i, (nums1, nums2, expected) in enumerate(test_cases, 1):
        result = solution.findMedianSortedArrays(nums1, nums2)
        status = "✓" if abs(result - expected) < 1e-5 else "✗"
        
        print(f"Test {i}: {status}")
        print(f"  Input: nums1 = {nums1}, nums2 = {nums2}")
        print(f"  Expected: {expected}")
        print(f"  Got: {result}")
        print()
    
    print("\nVerifying with brute force approach:")
    print("-" * 40)
    
    for i, (nums1, nums2, expected) in enumerate(test_cases[:5], 1):
        optimal_result = solution.findMedianSortedArrays(nums1, nums2)
        brute_result = solution.findMedianSortedArrays_bruteforce(nums1, nums2)
        
        match = abs(optimal_result - brute_result) < 1e-5
        status = "✓" if match else "✗"
        
        print(f"Test {i}: {status} Optimal: {optimal_result}, Brute: {brute_result}")


if __name__ == "__main__":
    test_solution()